pkg/model/project.go

```diff
@@ -243,3 +322,73 @@ func (p *ProjectSSOConfig_GitHub) GenerateAuthCodeURL(project, callbackURL, stat
 
 	return authURL, nil
 }
+
+// UpdateRBACRoles update only custom rbac roles.
+func (p *Project) UpdateRBACRoles(roles []*ProjectRBACRole) error {
+	v := make(map[string]struct{}, len(roles))
+	update := make([]*ProjectRBACRole, 0, len(roles))
+	for _, role := range roles {
+		if role.IsBuiltin {
+			continue
+		}
+		if role.IsBuiltinName() {
+			return fmt.Errorf("cannot use built-in role name")
+		}
+		if _, ok := v[role.Name]; ok {
+			return fmt.Errorf("role name must be unique")
+		}
+		v[role.Name] = struct{}{}
+		update = append(update, role)
+	}
+	p.RbacRoles = update
+	return nil
+}
+
+func (p *ProjectRBACRole) IsBuiltinName() bool {
+	return p.Name == builtinRBACRoleAdmin.String() ||
+		p.Name == builtinRBACRoleEditor.String() ||
+		p.Name == builtinRBACRoleViewer.String()
+}
+
+// UpdateUserGroups update user groups.
+func (p *Project) UpdateUserGroups(groups []*ProjectUserGroup) error {
+	v := make(map[string]struct{}, len(groups))
+	for _, group := range groups {
+		if _, ok := v[group.SsoGroup]; ok {
+			return fmt.Errorf("cannot assign multi roles to sso group")
+		}
+		v[group.SsoGroup] = struct{}{}
+	}
+	p.UserGroups = groups
+	return nil
+}
+
+// SetUserGroup add user group or update user group if already exists.
+func (p *Project) SetUserGroup(group, role string) {
+	v := &ProjectUserGroup{
+		SsoGroup: group,
+		Role:     role,
+	}
+	for i := range p.UserGroups {
+		if p.UserGroups[i].SsoGroup == group {
+			p.UserGroups[i] = v
+			return
+		}
+	}
+	p.UserGroups = append(p.UserGroups, v)
+}
+
+// MigrateFromRBAC migrate rbac config to user group.
+func (p *Project) MigrateFromRBAC() {
+	rbac := p.Rbac
+	if rbac.GetAdmin() != "" {
+		p.SetUserGroup(rbac.Admin, builtinRBACRoleAdmin.String())
+	}
+	if rbac.GetEditor() != "" {
+		p.SetUserGroup(rbac.Editor, builtinRBACRoleEditor.String())
+	}
+	if rbac.GetViewer() != "" {
+		p.SetUserGroup(rbac.Viewer, builtinRBACRoleViewer.String())
+	}
+	p.Rbac = &ProjectRBACConfig{}
+}
```

Comment:
Sorry, why do we need this migrate function?
I think at the last MTG, we finally decided to build the list of UserGroups based on the both new RBAC and old RBAC instead of migrating between them. 🤔